server/src/main/java/org/elasticsearch/health/GetHealthAction.java

```diff
@@ -0,0 +1,223 @@
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.health.components.controller.Controller;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class GetHealthAction extends ActionType<GetHealthAction.Response> {
+
+    public static final GetHealthAction INSTANCE = new GetHealthAction();
+    // TODO: Decide on name
+    public static final String NAME = "cluster:monitor/health/get";
+
+    private GetHealthAction() {
+        super(NAME, GetHealthAction.Response::new);
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private final ClusterName clusterName;
+        private final ClusterHealthStatus status;
+        private final int numberOfNodes;
+        private final int numberOfDataNodes;
+        private final String masterNodeId;
+        private final List<Component> components;
+
+        public Response(StreamInput in) {
+            throw new AssertionError("GetHealthAction should not be sent over the wire.");
+        }
+
+        public Response(
+            final ClusterName clusterName,
+            final int numberOfNodes,
+            final int numberOfDataNodes,
+            final String masterNodeId,
+            final List<Component> components
+        ) {
+            this.clusterName = clusterName;
+            this.numberOfNodes = numberOfNodes;
+            this.numberOfDataNodes = numberOfDataNodes;
+            this.masterNodeId = masterNodeId;
+            this.components = components;
+            ClusterHealthStatus computeStatus = ClusterHealthStatus.GREEN;
+            for (Component component : components) {
+                if (component.status().equals(ClusterHealthStatus.RED)) {
```

Comment:
Can we add a method to ClusterHealthStatus like `aggregate(ClusterHealthStatus` that does this logic?

I am slightly reluctant to reuse `ClusterHealthStatus` but do not really have a good argument for now. As long as we ensure it does not shine through the REST APIs.

Reply:
I'm also inclined to make our own enum here because I think we will want to diverge from `ClusterHealthStatus`. For instance, if the health subsystem is down (e.g. no master node, cannot locate persistent task) then how will we report the status of the components that need the full health subsystem to analyse? `GREEN` and `YELLOW` seem inappropriate, but reporting everything as `RED` is also unhelpful. I see value in having a new `UNKNOWN` status too.

Reply:
I made a basic enum. Do I need a custom serialization scheme? It is just green/yellow/red right now. Presumably a custom serialization scheme makes BWC explicit?

Reply:
It makes sense to me to have that be explicit.